### ISSUE TITLE
Separate Model from View

### ISSUE THREAD

**Issue by backspaces** [backspaces/asx#17]
The current AgentScript attempts to make the view and model separate, managed by the class Model module.

It succeeds as well as can be expected if the goal is to have swappable views, but it does so by having view variables in the model .. ex: shape, size, color, strokeColor, sprite, labels

AgentScript's main user, [SimTable](http://www.simtable.com/), has requested a more complete separation: a model-only, renderer-less, no-dependency "core" that is completely divorced from the model's view.

A "pure" solution would be a modeling core, with no knowledge of view, with views having complete knowledge of the core.

A less pure solution would be using a renderer-less class Model which allowed the existing agentsets (patches, turtles, links) to do no rendering via a "null renderer".

Further, a rollup of just the necessary modules is desired. This would exclude SpriteSheet, Color, ColorMap for example.

**Comment by stigmergic** [comment]
@backspaces: I'm still not convinced that these are "view" variables:  "shape, size, color, strokeColor, sprite, labels".

They seem to be properties of agents or patches.  

Are they in some way dependent on a View?

**Comment by backspaces** [comment]
TL;DR: Bingo! That why this is hard. I'm torn on this. 

Certainly label can be a "note" rather than visualized, so we certainly can keep it, I'll take it off the list. An issue, however, is how these are handled by the agentsets via getters/setters. 

Details: 
For example patches. When you set a patch (agent) color, it sets a pixel value in the patches (agentset). It uses a getter/setter pair to make this invisible to the programmer. But all views I know of use the pixels directly in their putting pixels on the screen. Either canvas2D drawImage or webgl Texture.

For links, the color is simply a property, no stunts.

For turtles, however, things get more difficult. For 2D canvases we use only sprites, and setting shape, size, color, strokeColor simply updates the sprite. We do this "lazily" by setting a needsUpdate flag for the renderer. We also have several turtles renderers. WebGL Points (thus no shape needed), either monochrome (no color needed) or colored plus your basic sprite (SpriteSheet with UV mapping).

But this is all easily simplified, and implemented in the existing Turtle class. Basically put shape, size, color, strokeColor into an object. These are all sprite attributes. If the model is monochrome, setting the color is a no-op with a one-time warning. If a Point, the shape is set to "rect". Size is used by the renderer in 2D & 3D when drawing the turtle.

So (sorry for all the details) these variables are often used by getter/setters to work well with the renderer and have been used by both 2D and 3D, canvas & webgl.

**Comment by stigmergic** [comment]
Ah, so these are privileged attributes of the model.  This ultimately seems like what we might need though. Perhaps in a more abstract nature.  Couldn't patches have more than one variable of type color? How are other patch variables implemented?

What I would expect would be that the view would be choosing how to render attributes,  while the model allows me to add attributes as needed (including potentially optimized types of variables useful for rendering).  To me the larger point is that we don't want the model to depend on communicating with the view.  Not that we are going to strictly maintain presentational variables separately from the model.  A Turtle might be rendered as a `<div>` where the turtle's color flows into its `style` in some way determined by the renderer.

All that being said,  we should trim out what we know doesn't feel right or is making choices based on the view -- we can always put things back in as the implementation becomes more clear.

**Comment by stephenguerin** [comment]
@stigmergic writes:
@backspaces: I'm still not convinced that these are "view" variables: "shape, size, color, strokeColor, sprite, labels".

As a mental exercise, consider 4 distributed views on 4 different machines which are being used by different roles in an emergency. One user will be setting color of turtles (people) based on if they are in communication or not. Another may color them based on their agency type (eg, USFS, BLM, Sheriff, Civilian etc). ditto for shape, size, sprite and labels. 

Netlogo modelers will often use these view properties in their model logic as a way for filtering (eg "ask turtles with [shape = "cirlcle" and color = green]") so I think that gives some confusion. To me that's a convenience hack to keep memory footprint down instead of introducing new attributes, not understanding agentSet variables or minimizing the need to introduce -own variables with beginners.

Or consider a ESRI shape file as a "model". Geometry is specified in the model (point, polyline, and polygon) but not markers (shape). Different GIS users will render the data model differently. KML files by contrast stuff the view into the data and is a counter-example of what we're trying to do.

PS. does issue have any way of threading discussions?

**Comment by backspaces** [comment]
> PS. does issue have any way of threading discussions?

Yes, by starting a new issue. No, within the same issue.

For example, we may decide this issue is too broad and break it down into separate ones.

After thinking about this, and talking to @stephenguerin & @stigmergic, a good first deliverable on this would be a way to have a modeler create renderer-less, with no Three.js dependencies.

We could make that a separate issue, or turn this issue into creating renderer-less models .. titled:
> Produce a means to create a Renderer-less Model

or something similar.

**Comment by backspaces** [comment]
Speaking of which, I'm going to start by looking at renderer-less by:
* Update class Model to have an option for no View.
  * I.e. remove the current connection to Three.js and the ThreeView and ThreeMeshes modules 
    https://github.com/backspaces/asx/tree/master/src
* A renderer-less Model would be signaled by the ctor having only the World object.
  * Current ctor has div, worldOptions, renderOptions. Renderer-less doesn't need div & renderOptions
* Yet another Rollup, sigh, would be created without ThreeView and ThreeMeshes.
* The existing AgentSet Modules would remain the same w.r.t. View variables
* One-time console warnings would be given for use of View variables
  * This already is done for ThreeMeshes being used with inappropriate variables.
* The existing test Models with ThreeView will remain as they currently are.

I realize this is just a first step and I'm sure I'll find "Devils in the Details" that may change some of the above.

I am concerned that introducing additional workflow will impact #14. Indeed, I'd much prefer we look at simplifying workflow by reducing the requirements on AS.

**Comment by backspaces** [comment]
Recent [changes to Rollup](https://github.com/rollup/rollup/issues/1169#issuecomment-331672466) addressed in [recent checkin](https://github.com/backspaces/asx/commit/3d4c3c6d60faa868116226ce0caa0526a0e01cb1).

This will help in rollups for "core". Workflow is a bitch.

**Comment by backspaces** [comment]
After looking at this in depth, it seems to me there are three basic approaches:
1. Modify the current Three.js system to be able to simply not draw.
    - Just give the Model ctor a div not in the DOM & hidden. Then make draw() a no-op
    - Requires SimTable to handle the Three.js dependency which has proven to be difficult.
    - Or use es6 Modules and let AS Modules refer to Three "[wrappers](https://medium.com/@backspaces/es6-modules-part-2-libs-wrap-em-up-8715e116d690)".
2. Create a second Module.js .. lets call it Module0.js, which eliminates the Three.js View
    - .. but keeps the View variables in Patch, Turtle, Link treated as they currently are.
    - .. requiring a "null-renderer" where existing Patch, Turtle, Link expect a renderer.
    - .. which is non-trivial but likely doable.
    - Hopefully minimally impacts the existing system but gives SimTable a Three.js-less rollup
3. Attempt removing all View variables and "special treatment" of them.
    - Minimally a Model0.js, Patch0.js, Turtle0.js, Link0.js module set managed by Module0.js
    - Current Module.js, Patch.js, Turtle.js, Link.js would subclass from the Foo0.js's
    - Very clean but "why me", adds significant workflow, and is a LOT of subtle work.

I don't know how realistic these are but after staring at the code, these seem the most reasonable. Pros/Cons:

1. Seems a bit silly, and still has the Three.js ton-of-code. Could be Rollup-ed.
2. Solves Three.js elimination but doesn't really add to AS.
3. Clearly cleanest, but adds a lot of workflow and confusing modules.

I do like the simplicity introduced by MV separation, but it's a lot of work while there are lot more improvements I'd like to make in the existing system. Also, we've all agreed that AS will have Views for simpler, NetLogo-like modeling researchers will be using, so is this really a high priority?

PR's welcome!  :)

**Comment by stigmergic** [comment]
Seems like option 2. is the best next step.  I am wondering what it would take to create a strictly canvas view.  Or even a view that just outputs JSON snapshots or the like.  I think making the model configurable so that views could be swapped out would be a win and an improvement for AS (from my very quick look at the code, it looks like this is already partially in place).

To beat a dead horse, I wouldn't worry too much about the view variables.  Ultimately a model has to communicate enough information to the view to be rendered,  having some default properties isn't the worst thing and it doesn't mean the view has to slavishly implement them.  As Steve points out it should be the renderer's prerogative to chose what properties to render and how.

**Comment by stigmergic** [comment]
That is maybe the model should use ThreeView as a default but allow for easily being configured to use a different view.  Then we could create a NOOP view (or even data snapshot view).

**Comment by backspaces** [comment]
I think I agree. Basically build new View ... maybe a DataView that could be configured to output nothing.

It is unfortunate that there is no "conditional import()" yet .. it's in the spec but not sure how far it has gone. I think we can play tricks with Rollup to eliminate modules we don't really use even tho imported.

Makes my head spin! But yes, number 2 it is.

**Comment by backspaces** [comment]
Looking at the View architecture, it looks like the simplest is to require the modeler to specify their View module. It makes the modeler's life a bit more annoying.

One way around this is to have multiple Model modules: Model (completely general, user specifies View) and then one per View type: ThreeModel, CanvasModel, DataModel, NullModel, etc. Hopefully identical in terms of code other than the FooModel naming and managing the Foo View. Each would have its own bundle, alas.

For now, I'm going to look into the modeler specifying the View, creating an instance, and directly setting any options it might have. We can look at simplifications later. We will still need a rollup per View, I think.

BTW: This is a case for es6 Modules rather than bundles/rollups. If the modeler simply imports individual modules, they don't drag in Three etc if they don't use it.

And if a bundle is needed, the absolute minimal bundle can be made of the specific Model itself!

**Comment by backspaces** [comment]
Let's drill down on option 4 above, "Create a new repo, for CoreAS, absolutely Viewless."

Here are the current modules. Those with a + would remain, but potentially quite simplified:
```
  + AS.js
  + AgentArray.js
  + AgentSet.js
    Animator.js
    AscDataSet.js
    Color.js
    ColorMap.js
  + DataSet.js
    DataSetIO.js
    Drawing.js
    Int24.js
  + Link.js
  + Links.js
  + Model.js
  + Patch.js
  + Patches.js
    RGBADataSet.js
    RGBDataSet.js
    SpriteSheet.js
    ThreeMeshes.js
    ThreeView.js
  + Turtle.js
  + Turtles.js
  + World.js
  + util.js
```
I do like how clean this is. The key difficulty is debugging without animation and a view. I'd like a reasonable strategy for this before starting.

**Comment by Vandivier** [comment]
Absolutely thrilled about this redesign. Will you be implementing as a NodeJS module?

I highly recommend that as performance will be off the charts, and able to leverage [C++ Add-ons.](https://nodejs.org/api/addons.html)

When using Node, the approach to debugging is to use loggers and logfiles. A 5 line logging function can be written which logs both to console/stdout and a specified log file using the built-in fs module. I don't recommend adding dependencies like Winston, Morgan, etc

If you remain client-oriented I think you would end up taking the same console.log() approach. You can write to file from the client as well, and you also have the debugger keyword. Actually, Node has debugger as well but I've never used it.

The eventual direction of the JS community is isomorphism. Have you taken a look at [.mjs modules](https://nodejs.org/api/esm.html)?

Another note is that the midst of this refactor may be a great time to introduce async.

Finally, how close to NetLogo will the library remain is an interesting question to me. Some module consolidation is possible eg Turtle is subsumed by Turtles, Link by Links, Patches by Patch and even Patch by World.

Depending on the scope of work we might want to leverage GitHub Projects, Trello, or some other tool in addition to Issues.

**Comment by backspaces** [comment]
Yup. I definitely plan on being node-capable for the model (not view) part. I'm a big fan of node and feel part of our strategy should include "agent lambdas" .. tiny Amazon Lambda distributed models.

The goal is to allow several participants in a larger model/app which would integrate multiple independent models and provide a view for the set of them.

The really hard part is how do you know if a model is working without being to "see" it! So maybe a fairly thin connection between a renderer and an agentset/agent. Or two separate class Model's that act as the integrator between the model and the view.

The ThreeView, for example, has a fairly separable architecture. The view is a collection of separate renderers, one per agentset.  In this case, the renderers are Three.js meshes which are children of the Scene. 

Each of the many meshes I have come with an Options object that expose their capabilities. The modeler choses the desired mesh and modifies the options object as needed.

So to explore this and find all the Devils in the Details, I've made a temporary repo of the modules with the "+" annotations above. Then I removed all use of color, size, shape so the resulting IIFE/es6 bundles will be minimal.

Fingers crossed! Toughest part is how to integrate the View with the Model.

**Comment by backspaces** [comment]
Update on list of modules: we're including a modified RGBDataSet using the MapBox elevation formula, slightly modified to be more general.

**Comment by backspaces** [comment]
The tests currently just loop n times. I’ve decided to put the animator back, however. Currently in Italy but when I get back in a few days.

**Comment by Vandivier** [comment]
I see. I made a SimpleBlindAnimator which may or may not be helpful. See [here](https://github.com/Vandivier/charm-abm/blob/master/src/static/scripts/simple-blind-anim.js).

I also use timeout({}, 0) instead of timeout({}, 20). is there a reason for those built in 20 ms delays?

Imo view-model separation means step()-draw() separation

**Comment by backspaces** [comment]
@Vandivier I like how clean this is. I'm planning on moving Animator back into core.

Maybe we need a few options like an event-driven animator?

**Comment by backspaces** [comment]
Just an update: I've made an initial model/view split. The model is a new repo:
https://github.com/backspaces/CoreAS

ASX uses this, with subclassing to add the current, somewhat clunky, ThreeView. 

I don't think subclassing is the way to go but it at least clarifies many of the issues.

I believe something more event based is what we're looking for. @stigmergic has used the core repo in a worker, thus requiring messaging .. therefore being more event oriented.
